File: preview3d/cinematics.py
"""Cinematic scene GENERATORS — a master rule for a whole family of scenes,
never one hand-authored descriptor per case (root Rule 19).

The Five Stations (PLAN.md, Extra 3D Views) "generalizes to any axis on
demand" — the cube has 13 of them, and hand-authoring 13 near-identical JSON
descriptors would be exactly the enumeration root Rule 19 forbids. This module
computes the scene descriptor for ANY axis of a model instead: the geometry
(vertex positions, the side-on camera angle, which OTHER parts to fade) is all
derived from the model's own data, so a new axis is a function argument, not a
new file.

Pure Python; the JS mirror is src/cinematics.js. Both read shared/spec.json for
the radial factors and switcher stop names, so neither restates them.
"""

from .directions import canonical_token, opposite_token, parse_direction
from .light.camera import Camera
from .model_scene import GROUP_PATHS, KIND_ORDER, TIER_ORDER
from .resources import load_shared_spec
from .vectors import basis_from, scale

_SPEC = load_shared_spec()
_TIERS: dict = _SPEC["axisTiers"]
_RADIAL: dict = _SPEC["switcher"]["radial"]
STOPS: list[str] = list(_SPEC["switcher"]["stops"])

FADE_END = 0.35     # the cube is gone by here
SLIDE_END = 0.75     # the beads have reached their stations by here
EASE = "ease-in-out"

# A gentle three-quarter opening pose — any axis starts here and eases to its
# own side-on angle, so no axis begins already looking straight down its line.
_OPENING = {"azimuth": 35.0, "elevation": 22.0}


def _find_axis(model: dict, axis_id: str) -> dict:
    wanted = canonical_token(axis_id)
    for axis in model["axes"]:
        if axis["id"] == wanted or opposite_token(axis["id"]) == wanted:
            return axis
    known = ", ".join(axis["id"] for axis in model["axes"])
    raise KeyError(f"Model {model['name']!r} has no axis {axis_id!r} — available: {known}")


def _side_on_angles(direction) -> tuple[float, float]:
    """Azimuth/elevation of a camera looking ACROSS `direction` rather than
    down it — the tertiary view's own trick (cube_model.py's DEFAULT_VIEWS),
    generalised to any axis instead of the one hardcoded perpendicular."""
    _, right, _ = basis_from(direction)
    camera = Camera()
    camera.look_along(right)
    return camera.azimuth, camera.elevation


def build_five_stations_scene(model: dict, axis_id: str, *, duration: float = 8.0) -> dict:
    """The Five Stations, played on ONE axis of `model`: the cube fades to a
    single line, the camera settles side-on, and the luminous/fallen beads
    slide from the geometric vertex to their radial stations (SCENES.md).
    """
    axis = _find_axis(model, axis_id)
    tier = axis["tier"]
    root = model.get("root") or "model"
    size = float(model.get("size", 1.0))
    length = float(_TIERS[tier]["length"]) * size
    axis_path = f"{root}/{GROUP_PATHS[tier]}/axis:{axis['id']}"
    centre_path = f"{root}/{GROUP_PATHS['centre']}"

    positive, negative = axis["ends"]
    azimuth, elevation = _side_on_angles(parse_direction(positive["direction"]))

    tracks = [
        _camera_track("camera.azimuth", _OPENING["azimuth"], azimuth),
        _camera_track("camera.elevation", _OPENING["elevation"], elevation),
        *_fade_everything_else(model, root, tier, axis["id"]),
        # Forced to 1 regardless of the view the content happened to open on —
        # this axis's own TIER group might have started dimmed (e.g. the
        # "cube" view leaves "secondary" at 0.5), and opacity multiplies down.
        _opacity_track(f"{root}/{GROUP_PATHS[tier]}", 1.0, 1.0),
        _opacity_track(axis_path, 1.0, 1.0),
        _opacity_track(centre_path, 1.0, 1.0),
    ]
    for end in (positive, negative):
        arm_path = f"{axis_path}/arm:{end['direction']}"
        direction = parse_direction(end["direction"])
        for stop in STOPS:
            stop_path = f"{arm_path}/{stop}"
            vertex_point = list(scale(direction, length))
            station_point = list(scale(direction, length * float(_RADIAL[stop])))
            tracks.append({
                "channel": "part.position", "path": stop_path,
                "keys": [
                    {"t": 0.0, "value": vertex_point},
                    {"t": FADE_END, "value": vertex_point, "ease": EASE},
                    {"t": SLIDE_END, "value": station_point},
                ],
            })
            tracks.append(_opacity_track(stop_path, 0.0, 1.0))

    return {
        "name": f"five_stations:{axis['id']}",
        "label": f"Five Stations — {axis['id']}",
        "duration": duration,
        "loop": False,
        "content": {"type": "model", "view": "cube"},
        "tracks": tracks,
    }


def _camera_track(channel: str, start: float, end: float) -> dict:
    return {
        "channel": channel,
        "keys": [{"t": 0.0, "value": start, "ease": EASE}, {"t": FADE_END, "value": end}],
    }


def _opacity_track(path: str, start: float, end: float) -> dict:
    return {
        "channel": "part.opacity", "path": path,
        "keys": [{"t": 0.0, "value": start, "ease": EASE}, {"t": SLIDE_END, "value": end}],
    }


def _fade_everything_else(model: dict, root: str, tier: str, axis_id: str) -> list[dict]:
    """Every group that is NOT this axis or the centre, faded to nothing by
    FADE_END — "the cube fades until only the Sacred Axis line remains"."""
    paths = [f"{root}/{GROUP_PATHS['glass']}"]
    for other_tier in TIER_ORDER:
        if other_tier != tier:
            paths.append(f"{root}/{GROUP_PATHS[other_tier]}")
    for sibling in model["axes"]:
        if sibling["tier"] == tier and sibling["id"] != axis_id:
            paths.append(f"{root}/{GROUP_PATHS[tier]}/axis:{sibling['id']}")
    for kind in KIND_ORDER:
        if kind != "centre":
            paths.append(f"{root}/{GROUP_PATHS[kind]}")
    return [_opacity_track(path, 1.0, 0.0) for path in paths]

File: src/cinematics.js
// Cinematic scene GENERATORS — a master rule for a whole family of scenes,
// never one hand-authored descriptor per case (root Rule 19).
//
// The Five Stations (PLAN.md, Extra 3D Views) "generalizes to any axis on
// demand" — the cube has 13 of them, and hand-authoring 13 near-identical JSON
// descriptors would be exactly the enumeration root Rule 19 forbids. This
// module computes the scene descriptor for ANY axis of a model instead: the
// geometry (vertex positions, the side-on camera angle, which OTHER parts to
// fade) is all derived from the model's own data, so a new axis is a function
// argument, not a new file.
//
// Mirror of preview3d/cinematics.py. Pure data — no three.js import, matching
// the rest of the model layer (directions.js, modelscene.js, switcher.js).

import SHARED from '../shared/spec.json';
import { canonicalToken, oppositeToken, parseDirection } from './directions.js';
import { GROUP_PATHS, KIND_ORDER, TIER_ORDER } from './modelscene.js';

const TIERS = SHARED.axisTiers;
const RADIAL = SHARED.switcher.radial;
export const STOPS = SHARED.switcher.stops;

const FADE_END = 0.35;     // the cube is gone by here
const SLIDE_END = 0.75;    // the beads have reached their stations by here
const EASE = 'ease-in-out';

// A gentle three-quarter opening pose — any axis starts here and eases to its
// own side-on angle, so no axis begins already looking straight down its line.
const OPENING = { azimuth: 35.0, elevation: 22.0 };

function findAxis(model, axisId) {
    const wanted = canonicalToken(axisId);
    const axis = model.axes.find((candidate) => candidate.id === wanted || oppositeToken(candidate.id) === wanted);
    if (!axis) {
        throw new Error(`Model '${model.name}' has no axis '${axisId}' — available: ${model.axes.map((a) => a.id).join(', ')}`);
    }
    return axis;
}

function cross(a, b) {
    return [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]];
}

function normalize(v) {
    const length = Math.hypot(...v);
    return v.map((c) => c / length);
}

function scale(v, k) {
    return v.map((c) => c * k);
}

// Azimuth/elevation of a camera looking ACROSS `direction` rather than down
// it — the tertiary view's own trick (cubemodel.js's DEFAULT_VIEWS),
// generalised to any axis instead of the one hardcoded perpendicular. A unit
// vector's asin(y) and atan2(y, hypot(x,z)) are the same angle exactly, which
// is what lets this use the shorter form while Python's Camera.look_along uses
// atan2 (it also accepts non-unit vectors) — one identity, not two formulas.
function sideOnAngles(direction) {
    const reference = Math.abs(direction[1]) > 0.999 ? [0, 0, 1] : [0, 1, 0];
    const right = normalize(cross(reference, direction));
    const [x, y, z] = right;
    const horizontal = Math.hypot(x, z);
    const azimuth = horizontal ? (Math.atan2(x, z) * 180) / Math.PI : 0;
    const elevation = horizontal ? (Math.asin(y) * 180) / Math.PI : (y > 0 ? 89.99 : -89.99);
    return [azimuth, elevation];
}

function cameraTrack(channel, start, end) {
    return { channel, keys: [{ t: 0.0, value: start, ease: EASE }, { t: FADE_END, value: end }] };
}

function opacityTrack(path, start, end) {
    return { channel: 'part.opacity', path, keys: [{ t: 0.0, value: start, ease: EASE }, { t: SLIDE_END, value: end }] };
}

// Every group that is NOT this axis or the centre, faded to nothing by
// FADE_END — "the cube fades until only the Sacred Axis line remains".
function fadeEverythingElse(model, root, tier, axisId) {
    const paths = [`${root}/${GROUP_PATHS.glass}`];
    for (const otherTier of TIER_ORDER) {
        if (otherTier !== tier) paths.push(`${root}/${GROUP_PATHS[otherTier]}`);
    }
    for (const sibling of model.axes) {
        if (sibling.tier === tier && sibling.id !== axisId) {
            paths.push(`${root}/${GROUP_PATHS[tier]}/axis:${sibling.id}`);
        }
    }
    for (const kind of KIND_ORDER) {
        if (kind !== 'centre') paths.push(`${root}/${GROUP_PATHS[kind]}`);
    }
    return paths.map((path) => opacityTrack(path, 1.0, 0.0));
}

// The Five Stations, played on ONE axis of `model`: the cube fades to a single
// line, the camera settles side-on, and the luminous/fallen beads slide from
// the geometric vertex to their radial stations (SCENES.md).
export function buildFiveStationsScene(model, axisId, duration = 8.0) {
    const axis = findAxis(model, axisId);
    const tier = axis.tier;
    const root = model.root || 'model';
    const size = Number(model.size ?? 1.0);
    const length = Number(TIERS[tier].length) * size;
    const axisPath = `${root}/${GROUP_PATHS[tier]}/axis:${axis.id}`;
    const centrePath = `${root}/${GROUP_PATHS.centre}`;

    const [positive, negative] = axis.ends;
    const [azimuth, elevation] = sideOnAngles(parseDirection(positive.direction));

    const tracks = [
        cameraTrack('camera.azimuth', OPENING.azimuth, azimuth),
        cameraTrack('camera.elevation', OPENING.elevation, elevation),
        ...fadeEverythingElse(model, root, tier, axis.id),
        // Forced to 1 regardless of the view the content happened to open on —
        // this axis's own TIER group might have started dimmed (e.g. the
        // "cube" view leaves "secondary" at 0.5), and opacity multiplies down.
        opacityTrack(`${root}/${GROUP_PATHS[tier]}`, 1.0, 1.0),
        opacityTrack(axisPath, 1.0, 1.0),
        opacityTrack(centrePath, 1.0, 1.0),
    ];
    for (const end of [positive, negative]) {
        const armPath = `${axisPath}/arm:${end.direction}`;
        const direction = parseDirection(end.direction);
        for (const stop of STOPS) {
            const stopPath = `${armPath}/${stop}`;
            const vertexPoint = scale(direction, length);
            const stationPoint = scale(direction, length * Number(RADIAL[stop]));
            tracks.push({
                channel: 'part.position', path: stopPath,
                keys: [
                    { t: 0.0, value: vertexPoint },
                    { t: FADE_END, value: vertexPoint, ease: EASE },
                    { t: SLIDE_END, value: stationPoint },
                ],
            });
            tracks.push(opacityTrack(stopPath, 0.0, 1.0));
        }
    }

    return {
        name: `five_stations:${axis.id}`,
        label: `Five Stations — ${axis.id}`,
        duration,
        loop: false,
        content: { type: 'model', view: 'cube' },
        tracks,
    };
}
